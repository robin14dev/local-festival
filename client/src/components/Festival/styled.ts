import styled from "styled-components";

export const Container = styled.article`
  width: 25%;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  cursor: pointer;

  & > img {
    max-width: 100%;
    aspect-ratio: 1/1;
    border-radius: 10px;
  }
  section {
    display: flex;
    justify-content: space-between;
    padding: 0 0.5rem;

    div {
      width: 85%;
    }

    h1 {
      font-weight: 600;
      font-size: 14px;
      line-height: 17px;
      margin: 0.4rem 0 0.18rem 0;
    }

    li {
      font-weight: 400;
      font-size: 14px;
      line-height: 17px;
      color: #797979;
    }

    h1,
    li {
      white-space: nowrap;
      overflow: hidden;
      width: 11rem;
      text-overflow: ellipsis;
    }
  }

  button {
    /* background-color: white; */
  }

  button > img {
    height: 1.5rem;
    width: auto;
  }

  transition: transform 0.3s ease-out;

  &:hover {
    transform: scale(1.03);
  }

  @media screen and (max-width: 1100px) {
    width: 33%;
  }
  @media screen and (max-width: 923px) {
    width: 50%;
  }
  @media screen and (max-width: 521px) {
    width: 100%;
  }
  @media (max-width: 485px) {
    padding: 1rem 0;
    &:hover {
      transform: none;
    }

    section {
      h1,
      li {
        width: 100%;
      }
    }
  }
`;

export const Status = styled.div<{ status: string }>`
  position: absolute;
  width: 52px;
  height: 26px;
  border-radius: 0.5rem 0.1rem 0.4rem 0.1rem;

  background-color: ${({ status }) =>
    status === "scheduled"
      ? `var(--primaryBlue)`
      : status === "completed"
        ? "#4e4d4d"
        : `var(--primaryOrange)`};
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
`;
