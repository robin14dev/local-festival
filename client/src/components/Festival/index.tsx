import React, { useContext, useEffect, useState } from "react";
import moment from "moment";
import { useNavigate } from "react-router-dom";
import { LoginModalDispatchContext } from "../../contexts/LoginModalContext";
import { PickItemsContext } from "../../contexts/PickItemsContext";
import { UserContext } from "../../contexts/userContext";
import { Container, Img, Section, Status } from "./styled";
import { NoImage, HeartIcon, EmptyHeartIcon } from "../../assets";
import "../../styles/common.scss";

type FestivalProps = {
  festival: FestivalItem;
};

export function getCurrentDate() {
  const today = new Date();
  const year = today.getFullYear();
  let month: string | number = today.getMonth() + 1;
  let day: string | number = today.getDate();

  month = month < 10 ? "0" + month : month;
  day = day < 10 ? "0" + day : day;

  return Number(`${year}${month}${day}`);
}
function getStatus(
  today: number,
  startDate: number,
  endDate: number
): [string, string] {
  const statusMap = {
    scheduled: "예정",
    completed: "종료",
    inProgress: "진행중",
  };

  const entries = Object.entries(statusMap);

  // 순서 중요
  if (startDate > today) {
    // 예정
    return entries[0];
  }
  if (today > endDate) {
    return entries[1];
  }
  return entries[2];
}

const Festival = ({ festival }: FestivalProps) => {
  const {
    authState: { loginStatus },
  } = useContext(UserContext);
  const setIsLoginModal = useContext(LoginModalDispatchContext);
  const { pickItems, togglePick } = useContext(PickItemsContext);
  const { festivalId, title, imageUrl, startDate, endDate, location } =
    festival;
  const [like, setLike] = useState(false);
  const navigate = useNavigate();

  const onErrorImg = (e: React.ChangeEvent<HTMLImageElement>) => {
    e.target.src = NoImage;
  };
  const onClickMoveDVP = (festivalId: number) => {
    navigate(`/Detail/${festivalId}`);
  };
  const onClickHandler = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    if (loginStatus) {
      togglePick(festival);
      setLike((prevLike) => !prevLike);
    } else {
      setIsLoginModal(true);
    }
  };
  const today = getCurrentDate();
  const [statusKey, statusText] = getStatus(today, startDate, endDate);

  useEffect(() => {
    if (loginStatus) {
      const isPicked = pickItems.some((pick) => pick.festivalId === festivalId);
      if (isPicked) {
        setLike(true);
      } else {
        like && setLike(false);
      }
    }
  }, [pickItems, loginStatus]);

  return (
    <Container
      data-testid="Festival"
      key={festivalId}
      onClick={() => onClickMoveDVP(festivalId)}
    >
      <Status status={statusKey}>{statusText}</Status>
      <Img
        src={imageUrl}
        data-testid="thumbnail"
        alt={title}
        onError={onErrorImg}
      />
      <Section>
        <div>
          <h1>{title}</h1>
          <ul>
            <li>{location}</li>
            <li>
              {moment(startDate, "YYYY.MM.DD").format("YYYY.MM.DD")} ~{" "}
              {moment(endDate, "YYYY.MM.DD").format("YYYY.MM.DD")}
            </li>
          </ul>
        </div>
        <button onClick={onClickHandler}>
          <img alt="heart" src={like ? HeartIcon : EmptyHeartIcon} />
        </button>
      </Section>
    </Container>
  );
};

export default Festival;
