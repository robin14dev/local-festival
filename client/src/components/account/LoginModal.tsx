import axios, { AxiosError } from "axios";
import React, { useCallback, useContext, useState } from "react";
import styled from "styled-components";
import Signup from "./Signup";
import { UserContext } from "../../contexts/userContext";

const Backdrop = styled.div<{ isHide: boolean }>`
  z-index: 20;
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100vh;
  background: rgba(0, 0, 0, 0.4);

  animation-duration: 0.4s;
  animation-name: ${(props) => (props.isHide ? "bright-soft" : "dark-soft")};
  animation-fill-mode: forwards;
`;
const Container = styled.div<{ isHide: boolean }>`
  z-index: 100;
  position: fixed;
  width: 400px;
  height: 460px;
  display: flex;
  flex-direction: column;

  text-align: center;
  background: #fff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.7);
  border-radius: 10px;
  margin-top: -256px;
  margin-left: -200px;
  left: 50%;
  top: 50%;

  animation-duration: 0.4s;
  animation-name: ${(props) => (props.isHide ? "slide-up" : "slide-down")};
  animation-fill-mode: forwards;

  & > h1 {
    margin: 2rem 0;
    font-size: 3rem;
    cursor: pointer;
    font-style: italic;
    font-family: "HS-Regular";
    color: var(--mainColor);
  }

  form {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 2rem 0;
    input {
      border-radius: 6px;
      border: 1.8px solid #b1aeae;
      padding: 0.5rem;
      width: 80%;
      height: 3rem;
      font-family: "NanumSquareRound";
      &::placeholder {
        opacity: 0.5;
      }
    }
    div {
      min-height: 1rem;
      color: red;
      font-size: 0.8rem;
      align-self: flex-start;
      margin-left: 3rem;
      margin-bottom: 1rem;
      padding-top: 0.2rem;
    }
    button {
      width: 80%;
      height: 3rem;
      margin-top: 1rem;
      font-size: 1rem;

      color: #fff;
      font-family: "NanumSquareRound";
      font-weight: bold;
      background: var(--mainColor);
      border-radius: 6px;
      cursor: pointer;
    }
  }

  .footer {
    display: flex;
    /* background-color: yellow; */
    justify-content: center;
    align-items: center;
    font-size: 0.9rem;
    font-weight: 500;

    button {
      font-family: "NanumSquareRound";
      margin-left: 1rem;
      font-size: 1rem;
      /* background-color: yellowgreen; */
      color: var(--primaryPurple);
      font-weight: bold;
      padding-bottom: 3px;
      transition: all 0.2s;
      &:hover {
        color: var(--primaryOrange);
        border-radius: 5px;
      }
    }
  }

  p {
    color: red;
  }

  .modalClose {
    display: none;
  }

  @media screen and (max-width: 428px) {
    width: 100%;
    height: 100%;
    border-radius: 0;
    margin: 0;
    left: 0;
    top: 0;

    .modalClose {
      display: block;
      margin-top: 2rem;
      color: gray;
      font-family: "NanumSquareRound";
    }
  }
`;

/*
백드롭은 계속 어둡게 하고 
로그인  => 회원가입 => 성공 => 실패

*/

type LoginProps = {
  setIsLoginModal: React.Dispatch<React.SetStateAction<boolean>>;
};

const LoginModal = ({ setIsLoginModal }: LoginProps) => {
  console.log("logimModal!!!");

  const { setAuthState } = useContext(UserContext);
  const [userInfo, setUserInfo] = useState({ account: "", password: "" });
  const { account, password } = userInfo;
  const [errMessage, setErrMessage] = useState("");
  const [isHide, setIsHide] = useState(false);
  const [isLogin, setIsLogin] = useState(true);
  const [isSignup, setIsSignup] = useState(false);

  const errMessages = [
    "사용자가 존재하지 않습니다",
    "비밀번호가 일치하지 않습니다",
  ];

  const handleUserInfo = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      if (e.target.value.length > 0) setErrMessage("");
      setUserInfo((prevInfo) => ({
        ...prevInfo,
        [e.target.name]: e.target.value,
      }));
    },
    []
  );
  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    try {
      e.preventDefault();
      const response = await axios.post(
        `${process.env.REACT_APP_SERVER_URL}/users/signin`,
        {
          account: userInfo.account,
          password: userInfo.password,
        }
      );
      const { nickname, userId, account, defaultPic } = response.data.info;
      const user = {
        account,
        nickname,
        userId,
        defaultPic,
        loginStatus: true,
      };
      setAuthState(user);
      sessionStorage.setItem("user", JSON.stringify(user));
      sessionStorage.setItem("accesstoken", response.data.info.token);

      /**
       * loginHandler 안에 로그인 성공시, 찜 데이터를 가져 와야 하는데 분리해 주어함
       *
       * 로그인버튼 누른다
       * 로그인을 성공해서 유저정보를 가져온다.
       * 가져온 유저정보로 authState를 업데이트 한다
       * 해당 유저정보로 다시 찜 정보를 가져오기 위해 pick api를 보낸다.
       */

      modalClose();
    } catch (err: unknown) {
      if (err instanceof AxiosError) {
        console.log(err.response?.data.message);
        if (
          err.response?.data.message ===
          "Wrong account And Password Combination"
        ) {
          setErrMessage(errMessages[1]);
        } else if (err.response?.data.message === "User Doesn't Exist") {
          setErrMessage(errMessages[0]);
        } else {
          console.log("그밖에에러");
        }
      } else {
        console.log(err);
      }
    }
  };
  const modalClose = () => {
    setIsHide(true);
    setIsSignup(false);
    setTimeout(() => {
      setIsLoginModal(false);
    }, 400);
  };
  return (
    <>
      <Backdrop onClick={modalClose} isHide={isHide}>
        {isLogin && (
          <Container
            data-testid="LoginModal"
            isHide={isHide}
            onClick={(e) => {
              e.stopPropagation();
            }}
          >
            <h1>Loco</h1>
            <form onSubmit={handleSubmit}>
              <input
                name="account"
                placeholder="email을 입력해 주세요"
                onChange={handleUserInfo}
                type="text"
                value={account}
                required
              ></input>
              <div>
                {errMessage === "사용자가 존재하지 않습니다" && errMessage}
              </div>
              <input
                name="password"
                placeholder="비밀번호를 입력해 주세요"
                onChange={handleUserInfo}
                type="password"
                value={password}
                required
              ></input>
              <div>
                {errMessage === "비밀번호가 일치하지 않습니다" && errMessage}
              </div>
              <button type="submit">로그인</button>
            </form>
            <div className="footer">
              <span>아직 계정이 없으신가요?</span>
              <button
                onClick={() => {
                  setIsLogin(false);
                  setIsSignup(true);
                }}
              >
                회원가입
              </button>
            </div>
            <button
              className="modalClose"
              onClick={() => {
                setIsLoginModal(false);
              }}
            >
              돌아가기
            </button>
          </Container>
        )}

        {isSignup && (
          <Signup setIsSignup={setIsSignup} setIsLogin={setIsLogin} />
        )}
      </Backdrop>
    </>
  );
};

export default LoginModal;
