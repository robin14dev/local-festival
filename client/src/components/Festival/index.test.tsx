import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import React, { useContext } from "react";
import Festival from ".";
import {
  mockAuthState,
  mockFestivalItem,
  mockFestivalItems,
} from "../../test/mocks/mockData";
import { MemoryRouter, Route, Routes, useLocation } from "react-router-dom";
import DetailView from "../../pages/DetailView";
import {
  LoginModalContextProvider,
  LoginModalStateContext,
} from "../../contexts/LoginModalContext";
import { UserContext, UserContextProvider } from "../../contexts/userContext";
import LoginModal from "../account/LoginModal";
import {
  PickItemsContext,
  PickItemsContextProvider,
} from "../../contexts/PickItemsContext";
import {
  interceptAddPickItem,
  interceptDeletePickItem,
  interceptGetPickItems,
} from "../../test/mocks/mockServer/scope/pick";
import userEvent from "@testing-library/user-event";
import { interceptGetFestivalDetail } from "../../test/mocks/mockServer/scope/festival";
import { interceptGetReviews } from "../../test/mocks/mockServer/scope/review";
import { NoImage } from "../../assets";

afterEach(() => {
  window.sessionStorage.clear();
});

// TODO :  api 실패했을 때도 추가해 줘야됨
/**
 * 토큰 없는 경우 401
 *
 * 서버 에러 503
 *
 * 추가할 테스트는 찜 추가 해제할 때
 */

describe("<Festival/>", () => {
  test("컴포넌트가 성공적으로 렌더링 됩니다", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/"]}>
        <Festival festival={mockFestivalItem} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
  test("오늘 날짜를 기준으로, 축제 일정에 따라 [진행중], [예정], [종료] 상태를 나타내는 UI 가 정상적으로 렌더링 됩니다.", () => {
    const testCases = [
      { festival: mockFestivalItems.completed, expectedText: "종료" },
      { festival: mockFestivalItems.inProgress, expectedText: "진행중" },
      { festival: mockFestivalItems.scheduled, expectedText: "예정" },
    ];
    testCases.forEach(({ festival, expectedText }) => {
      render(
        <MemoryRouter initialEntries={["/"]}>
          <Festival festival={festival} />
        </MemoryRouter>
      );
      expect(screen.getByText(expectedText)).toBeInTheDocument();
      cleanup();
    });
  });
  test("축제 썸네일 이미지가 없거나 유효하지 않은 url이면 기본 로고 이미지가 렌더링 됩니다.", () => {
    const validUrl =
      "https://images.unsplash.com/photo-1707343843344-011332037abb?q=80&w=2070&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D";
    const testCases = [
      { festival: { ...mockFestivalItem, imageUrl: "" }, src: NoImage },
      {
        festival: { ...mockFestivalItem, imageUrl: "invalidUrl" },
        src: NoImage,
      },
      { festival: mockFestivalItem, src: validUrl },
    ];

    for (const { festival } of testCases) {
      const { container } = render(
        <MemoryRouter initialEntries={["/"]}>
          <Festival festival={festival} />
        </MemoryRouter>
      );

      const imgElement = screen.getByTestId("thumbnail");

      if (["", "invalidUrl"].includes(festival.imageUrl)) {
        fireEvent.error(imgElement);
        expect(imgElement).toHaveAttribute("src", NoImage);
      } else {
        fireEvent.load(imgElement);
        expect(imgElement).toHaveAttribute("src", validUrl);
      }
      expect(container).toMatchSnapshot();
      cleanup(); // 얘를 해줘야됨!!!
    }

    // testCases.forEach(({ festival }, i) => {
    //   const { container } = render(
    //     <MemoryRouter initialEntries={["/"]}>
    //       <Festival festival={festival} />
    //     </MemoryRouter>
    //   );
    //   console.log("TEST", i);
    //   console.dir(container);

    //   const imgElement = screen.getByTestId("thumbnail");

    //   if (festival.imageUrl === "") {
    //     fireEvent.error(imgElement);
    //     expect(imgElement).toHaveAttribute("src", NoImage);
    //   } else {
    //     fireEvent.load(imgElement);
    //     expect(imgElement).toHaveAttribute("src", validUrl);
    //   }
    // expect(container).toMatchSnapshot()

    // const { container } = render(
    //   <MemoryRouter initialEntries={["/"]}>
    //     <Festival festival={{ ...mockFestivalItem, imageUrl: "notvalidUrl" }} />
    //   </MemoryRouter>
    // );

    // const imgElement = screen.getByAltText(mockFestivalItem.title);
    // fireEvent.error(imgElement);
    // expect(imgElement).toHaveAttribute("src", NoImage);
    // expect(container).toMatchSnapshot();
  });
  test("컴포넌트 클릭시 상세페이지로 이동합니다", async () => {
    const interceptAPIOnDetailPage = () => {
      interceptGetFestivalDetail();
      interceptGetReviews();
    };
    const TestPageComponent = (): JSX.Element => {
      return <Festival festival={mockFestivalItem} />;
    };
    const UrlCheck = (): JSX.Element => {
      const { pathname } = useLocation();
      return <div>pathname is {pathname}</div>;
    };
    render(
      <MemoryRouter initialEntries={["/Test"]}>
        <UrlCheck />
        <Routes>
          <Route path="/Test" element={<TestPageComponent />} />
          <Route path="/Detail/:festivalId/*" element={<DetailView />} />
        </Routes>
      </MemoryRouter>
    );
    const pathname = screen.getByText("pathname is /Test");
    expect(pathname).toBeInTheDocument();
    interceptAPIOnDetailPage();
    userEvent.click(screen.getByTestId("Festival"));

    waitFor(async () => {
      expect(pathname.textContent).toBe("pathname is /Detail/1");
      const loading = await screen.findByText("상세페이지를 불러오고 있습니다");
      expect(loading).toBeInTheDocument();
    });
  });
  test("로그인 상태시, 하트를 클릭하면 해당 축제가 찜 추가 / 찜 해제가 됩니다", async () => {
    const testCases = [
      {
        type: "addPick",
        pickItemsOnStorage: JSON.stringify([]),
        numOfInitialPickItems: 0,
        numOfitemsAfterClick: 1,
      },
      {
        type: "deletePick",
        pickItemsOnStorage: JSON.stringify([mockFestivalItem]),
        numOfInitialPickItems: 1,
        numOfitemsAfterClick: 0,
      },
    ];

    const MockApp = () => {
      const {
        authState: { loginStatus },
      } = useContext(UserContext);
      const { pickItems } = useContext(PickItemsContext);
      return (
        <>
          <div>numOfPickItems is {pickItems.length}</div>
          {loginStatus ? "login" : "logout"}

          <Festival festival={mockFestivalItem} />
        </>
      );
    };

    window.sessionStorage.setItem("user", JSON.stringify(mockAuthState.login));
    render(
      <MemoryRouter initialEntries={["/"]}>
        <PickItemsContextProvider>
          <UserContextProvider>
            <MockApp />
          </UserContextProvider>
        </PickItemsContextProvider>
      </MemoryRouter>
    );

    const loginStatus = await screen.findByText("login");
    expect(loginStatus).toBeInTheDocument();
    for (const {
      type,
      pickItemsOnStorage,
      numOfInitialPickItems,
      numOfitemsAfterClick,
    } of testCases) {
      window.sessionStorage.setItem("picks", pickItemsOnStorage);
      const numOfPickItems = await screen.findByText(
        `numOfPickItems is ${numOfInitialPickItems}`
      );
      expect(JSON.parse(sessionStorage.getItem("picks") as string).length).toBe(
        numOfInitialPickItems
      );
      expect(numOfPickItems).toBeInTheDocument();
      const heart = screen.getByAltText("heart");
      if (type === "addPick") {
        interceptAddPickItem();
      } else if (type === "deletePick") {
        interceptDeletePickItem();
      }
      userEvent.click(heart);
      const element = await screen.findByText(
        `numOfPickItems is ${numOfitemsAfterClick}`
      );
      expect(element).toBeInTheDocument();
    }
  });
  test("로그인 상태시, 하트를 클릭하면 해당 축제가 찜 목록에 추가됩니다", async () => {
    //# 찜을 추가하는 경우
    window.sessionStorage.setItem("user", JSON.stringify(mockAuthState.login));
    window.sessionStorage.setItem("picks", JSON.stringify([]));

    const MockApp = () => {
      const {
        authState: { loginStatus },
      } = useContext(UserContext);
      const { pickItems } = useContext(PickItemsContext);
      return (
        <>
          <div>numOfPickItems is {pickItems.length}</div>
          {loginStatus ? "login" : "logout"}

          <Festival festival={mockFestivalItem} />
        </>
      );
    };

    interceptGetPickItems();

    render(
      <MemoryRouter initialEntries={["/"]}>
        <PickItemsContextProvider>
          <UserContextProvider>
            <MockApp />
          </UserContextProvider>
        </PickItemsContextProvider>
      </MemoryRouter>
    );

    const loginStatus = await screen.findByText("login");
    expect(loginStatus).toBeInTheDocument();
    const numOfPickItems = await screen.findByText("numOfPickItems is 0");
    expect(numOfPickItems).toBeInTheDocument();
    const heart = screen.getByAltText("heart");

    interceptAddPickItem();

    userEvent.click(heart);

    const element = await screen.findByText("numOfPickItems is 1");
    expect(element).toBeInTheDocument();
  });
  test("로그인 상태시, 이미 찜한 축제의 하트를 클릭하면 찜 목록에서 해제됩니다", async () => {
    //# 찜을 해제 하는 경우
    window.sessionStorage.setItem("user", JSON.stringify(mockAuthState.login));
    window.sessionStorage.setItem("picks", JSON.stringify([mockFestivalItem]));

    const MockApp = () => {
      const {
        authState: { loginStatus },
      } = useContext(UserContext);
      const { pickItems } = useContext(PickItemsContext);
      return (
        <>
          <div>numOfPickItems is {pickItems.length}</div>
          {loginStatus ? "login" : "logout"}

          {pickItems.length && <Festival festival={pickItems[0]} />}
        </>
      );
    };

    render(
      <MemoryRouter initialEntries={["/"]}>
        <UserContextProvider>
          <PickItemsContextProvider>
            <MockApp />
          </PickItemsContextProvider>
        </UserContextProvider>
      </MemoryRouter>
    );

    expect(JSON.parse(sessionStorage.getItem("picks") as string).length).toBe(
      1
    );
    const loginStatus = await screen.findByText("login");
    expect(loginStatus).toBeInTheDocument();
    const numOfPickItems = await screen.findByText("numOfPickItems is 1");
    expect(numOfPickItems).toBeInTheDocument();
    const heart = screen.getByAltText("heart");

    interceptDeletePickItem();

    userEvent.click(heart);

    const element = await screen.findByText("numOfPickItems is 0");
    expect(element).toBeInTheDocument();
  });
  test("비 로그인 상태시, 하트를 클릭하면 로그인 모달이 나타납니다", () => {
    const MockApp = () => {
      const {
        authState: { loginStatus },
      } = useContext(UserContext);
      const isLoginModal = useContext(LoginModalStateContext);
      const { pathname } = useLocation();
      return (
        <>
          <div>pathname is {pathname}</div>
          {loginStatus ? "login" : "logout"}
          {isLoginModal && <LoginModal />}
          <Festival festival={mockFestivalItem} />
        </>
      );
    };
    const { container } = render(
      <MemoryRouter initialEntries={["/"]}>
        <LoginModalContextProvider>
          <UserContextProvider>
            <MockApp />
          </UserContextProvider>
        </LoginModalContextProvider>
      </MemoryRouter>
    );
    expect(sessionStorage.getItem("user")).toBeNull();
    const pathname = screen.getByText("pathname is /");
    expect(pathname).toBeInTheDocument();
    const loginStatus = screen.getByText("logout");
    expect(loginStatus).toBeInTheDocument();
    const loginModal = screen.queryByTestId("LoginModal");
    expect(loginModal).toBeNull();
    const heart = screen.getByAltText("heart");
    fireEvent.click(heart);
    const renderedLoginModal = screen.getByTestId("LoginModal");
    expect(renderedLoginModal).toBeInTheDocument();
    expect(pathname.textContent).toBe("pathname is /");
    expect(container).toMatchSnapshot();
  });
});
