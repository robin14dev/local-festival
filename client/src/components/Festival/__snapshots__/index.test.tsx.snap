// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Festival/> 비 로그인 상태시, 하트를 클릭하면 로그인 모달이 나타납니다 1`] = `
<div>
  <div>
    pathname is 
    /
  </div>
  logout
  <div
    class="sc-gYbzsP izQiId"
  >
    <div
      class="sc-cCjUiG hnfQcp"
      data-testid="LoginModal"
    >
      <h1>
        Loco
      </h1>
      <form>
        <input
          name="account"
          placeholder="email을 입력해 주세요"
          required=""
          type="text"
          value=""
        />
        <div />
        <input
          name="password"
          placeholder="비밀번호를 입력해 주세요"
          required=""
          type="password"
          value=""
        />
        <div />
        <button
          type="submit"
        >
          로그인
        </button>
      </form>
      <div
        class="footer"
      >
        <span>
          아직 계정이 없으신가요?
        </span>
        <button>
          회원가입
        </button>
      </div>
      <button
        class="modalClose"
      >
        돌아가기
      </button>
    </div>
  </div>
  <article
    class="sc-bcXHqe kuAQdD"
    data-testid="Festival"
  >
    <div
      class="sc-dkrFOg bqYawA"
    >
      진행중
    </div>
    <img
      alt="축제 제목"
      class="sc-hLBbgP gUqOWE"
      data-testid="thumbnail"
      src="https://images.unsplash.com/photo-1707343843344-011332037abb?q=80&w=2070&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
    />
    <section
      class="sc-gswNZR dVBOQb"
    >
      <div>
        <h1>
          축제 제목
        </h1>
        <ul>
          <li>
            위치 정보
          </li>
          <li>
            2024.01.01
             ~
             
            2024.12.31
          </li>
        </ul>
      </div>
      <button>
        <img
          alt="heart"
          src="emptyHeartIcon.png"
        />
      </button>
    </section>
  </article>
</div>
`;

exports[`<Festival/> 축제 썸네일 이미지가 없거나 유효하지 않은 url이면 기본 로고 이미지가 렌더링 됩니다. 1`] = `
<div>
  <article
    class="sc-bcXHqe kuAQdD"
    data-testid="Festival"
  >
    <div
      class="sc-dkrFOg bqYawA"
    >
      진행중
    </div>
    <img
      alt="축제 제목"
      class="sc-hLBbgP gUqOWE"
      data-testid="thumbnail"
      src="noImage.png"
    />
    <section
      class="sc-gswNZR dVBOQb"
    >
      <div>
        <h1>
          축제 제목
        </h1>
        <ul>
          <li>
            위치 정보
          </li>
          <li>
            2024.01.01
             ~
             
            2024.12.31
          </li>
        </ul>
      </div>
      <button>
        <img
          alt="heart"
          src="emptyHeartIcon.png"
        />
      </button>
    </section>
  </article>
</div>
`;

exports[`<Festival/> 축제 썸네일 이미지가 없거나 유효하지 않은 url이면 기본 로고 이미지가 렌더링 됩니다. 2`] = `
<div>
  <article
    class="sc-bcXHqe kuAQdD"
    data-testid="Festival"
  >
    <div
      class="sc-dkrFOg bqYawA"
    >
      진행중
    </div>
    <img
      alt="축제 제목"
      class="sc-hLBbgP gUqOWE"
      data-testid="thumbnail"
      src="noImage.png"
    />
    <section
      class="sc-gswNZR dVBOQb"
    >
      <div>
        <h1>
          축제 제목
        </h1>
        <ul>
          <li>
            위치 정보
          </li>
          <li>
            2024.01.01
             ~
             
            2024.12.31
          </li>
        </ul>
      </div>
      <button>
        <img
          alt="heart"
          src="emptyHeartIcon.png"
        />
      </button>
    </section>
  </article>
</div>
`;

exports[`<Festival/> 축제 썸네일 이미지가 없거나 유효하지 않은 url이면 기본 로고 이미지가 렌더링 됩니다. 3`] = `
<div>
  <article
    class="sc-bcXHqe kuAQdD"
    data-testid="Festival"
  >
    <div
      class="sc-dkrFOg bqYawA"
    >
      진행중
    </div>
    <img
      alt="축제 제목"
      class="sc-hLBbgP gUqOWE"
      data-testid="thumbnail"
      src="https://images.unsplash.com/photo-1707343843344-011332037abb?q=80&w=2070&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
    />
    <section
      class="sc-gswNZR dVBOQb"
    >
      <div>
        <h1>
          축제 제목
        </h1>
        <ul>
          <li>
            위치 정보
          </li>
          <li>
            2024.01.01
             ~
             
            2024.12.31
          </li>
        </ul>
      </div>
      <button>
        <img
          alt="heart"
          src="emptyHeartIcon.png"
        />
      </button>
    </section>
  </article>
</div>
`;

exports[`<Festival/> 컴포넌트가 성공적으로 렌더링 됩니다 1`] = `
<div>
  <article
    class="sc-bcXHqe kuAQdD"
    data-testid="Festival"
  >
    <div
      class="sc-dkrFOg bqYawA"
    >
      진행중
    </div>
    <img
      alt="축제 제목"
      class="sc-hLBbgP gUqOWE"
      data-testid="thumbnail"
      src="https://images.unsplash.com/photo-1707343843344-011332037abb?q=80&w=2070&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDF8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
    />
    <section
      class="sc-gswNZR dVBOQb"
    >
      <div>
        <h1>
          축제 제목
        </h1>
        <ul>
          <li>
            위치 정보
          </li>
          <li>
            2024.01.01
             ~
             
            2024.12.31
          </li>
        </ul>
      </div>
      <button>
        <img
          alt="heart"
          src="emptyHeartIcon.png"
        />
      </button>
    </section>
  </article>
</div>
`;
